tc/autotuner/options_cache.h

```diff
@@ -0,0 +1,180 @@
+
+#include "tc/core/compilation_cache.h"
+#include "tc/core/tensor.h"
+#include "tc/core/utils/time.h"
+#include "tc/lang/canonicalize.h"
+
+namespace tc {
+namespace autotune {
+
+/**
+ * A key in the options cache is exactly the content of the underlying proto in
+ * tc/proto/compcache.proto. It provides simple conversions and equality
+ * operator. Additionally we provide a hash function to allow it to be a key
+ * in a hash map.
+ */
+struct OptionsCacheKey {
+  std::string id;
+  std::vector<TensorInfo> inputs;
+  std::vector<TensorInfo> outputs;
+  std::string deviceStr;
+
+  inline bool operator==(const OptionsCacheKey& other) const;
+  inline OptionsCacheKeyProto toProtobuf() const;
+  static inline OptionsCacheKey fromProto(const OptionsCacheKeyProto& proto);
+};
+
+struct OptionsCacheKeyHash {
+  inline std::size_t operator()(const OptionsCacheKey& k) const;
+};
+
+/**
+ * A value in the options cache is exactly the content of the underlying proto
+ * in tc/proto/compcache.proto. An OptionsCacheValue is templated by the Backend
+ * and contains an object of the proper MappingOptions type as well as the
+ * runtimes of subsequent runs.
+ */
+template <typename Backend>
+struct OptionsCacheValue {
+  typename Backend::OptionsCacheValueProtoType toProtobuf() const;
+  static OptionsCacheValue<Backend> fromProto(
+      const typename Backend::OptionsCacheValueProtoType& proto);
+
+  std::vector<Duration> runtimes;
+  typename Backend::MappingOptionsType mappingOptions;
+};
+
+/**
+ * An Options cache is a simple abstraction around an unordered_multimap of
+ * protobuf-backed key value pairs. It has an underlying store object and
+ * provide simple functions to load/store from proto, record, prune and
+ * extract topK values ordered by runtime.
+ * An OptionsCache is templated by the backend type because the values stored
+ * are backend-dependent.
+ */
+template <typename Backend>
+struct OptionsCache {
+ public:
+  using KeyType = OptionsCacheKey;
+  using ValueType = OptionsCacheValue<Backend>;
+  using MultiMapType = std::unordered_multimap<
+      OptionsCacheKey,
+      OptionsCacheValue<Backend>,
+      OptionsCacheKeyHash>;
+
+  OptionsCache<Backend>();
+
+  /// Calls the underlying unordered_multimap::clear and resets the counters
+  void clear();
+
+  /// Calls the underlying unordered_multimap::count for a particular key
+  size_t count(const OptionsCacheKey& key) const;
+
+  //// Calls the underlying unordered_multimap::size
+  size_t size() const;
+
+  /// Collects the keys in the cache
+  /// \return an unordered_set of keys
+  std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash> getKeys() const;
+
+  /// Loads in place from proto file. Calls fillFromProto which can insert
+  /// duplicates, so be sure your cache is cleared if you don't want those
+  void loadCacheFromProtobuf(const std::string& filename);
+
+  /// Stores to a proto file at the specified location
+  void dumpCacheToProtobuf(const std::string& filename);
```

Comment:
> Can protobuf be serialized to some other type of stream instead of a file, network stream, string

I would hold that off until we do it consistently for the compilation caches (hint, at a later date).